File: AMS-core/ams/routes/leave.py
from fastapi import APIRouter, HTTPException, Query, status
from ams.schemas.leave import Leave, LeaveApproval
from ams.services.leave_services import leave_mgr


router = APIRouter()


@router.get("/student")
async def get_leave_by_student(student_id: str):
    try:
        leave_data = await leave_mgr.get_leave_by_student(student_id)
        return leave_data
    except Exception as e:
        raise e


@router.get("/faculty")
async def get_leave_by_faculty(user_id: str = Query(...)):
    try:
        leave_data = await leave_mgr.get_leave_by_faculty(user_id)
        return leave_data
    except Exception as e:
        raise e


@router.post("/submit-leave/")
async def submit_leave(leave: Leave):
    try:
        result = await leave_mgr.submit_leave(leave)
        if not result:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=f"Error Submitting Leave")
        return {"message": "Leave submitted successfully!"}
    except HTTPException as e:
        raise e
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                            detail=f"Error Submitting Leave {str(e)}")


@router.patch("/approve-leave/")
async def approve_leave(data: LeaveApproval):
    try:
        result = await leave_mgr.approve_leave(data)
        if not result:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=f"Error Approving Leave")
        return {"message": "Leave approved successfully!"}
    except HTTPException as e:
        raise e
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                            detail=f"Error Approving Leave {str(e)}")


@router.get("/{leave_id}/")
async def get_leave_by_id(leave_id: str):
    try:
        leave_data = await leave_mgr.get_leave_by_id(leave_id)
        if not leave_data:
            raise HTTPException(status_code=404, detail="Leave not found")
        return leave_data
    except HTTPException as e:
        raise e

File: AMS-ui/src/services/api/leave.ts
import axios from "axios";

const API_URL = process.env.NEXT_PUBLIC_BACKEND_URL

export const fetchLeave = async (student_id: string) => {
    try {
        const response = await axios.get(`${API_URL}/leave/student`, {
            params: { student_id },
        })
        return response.data
    } catch (error: any) {
        throw Error(error.message)
    }
}

export const fetchLeaveSubmitted = async (faculty_id: string) => {
    try {
        const response = await axios.get(`${API_URL}/leave/faculty/`, {
            params: { faculty_id },
        })
        return response.data
    } catch (error: any) {
        throw Error(error.message)
    }
}

export const submitLeave = async (leaveData: any) => {
    try {
        const response = await axios.post(`${API_URL}/leave/submit-leave/`, leaveData)
        return response.data
    } catch (error: any) {
        throw Error(error.message)
    }
}

export const approveLeave = async (leave_id: string, leave_status: string) => {
    try {
        const response = await axios.patch(`${API_URL}/leave/approve-leave/`, {
            leave_id,
            leave_status,
        })
        return response.data
    } catch (error: any) {
        throw Error(error.message)
    }
}

export const fetchLeaveById = async (leave_id: string) => {
    try {
        const response = await axios.get(`${API_URL}/leave/${leave_id}/`)
        return response.data
    } catch (error: any) {
        throw Error(error.message)
    }
}

export const getLeavesForFaculty = async (user_id: string) => {
    try {
        const reponse = await axios.get(`${API_URL}/leave/faculty`, {
            params: { user_id }
        })
        return reponse.data
    } catch (error: any) {
        throw Error(error.message)
    }
}